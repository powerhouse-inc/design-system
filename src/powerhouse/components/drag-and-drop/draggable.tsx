import { useDrag } from 'react-aria';
import { CUSTOM_OBJECT_FORMAT } from './constants';

export interface DraggableRenderProps {
    isDragging: boolean;
}

export interface DraggableProps<Item = any>
    extends Omit<React.HTMLAttributes<HTMLDivElement>, 'children'> {
    children?: (props: DraggableRenderProps) => React.ReactNode;
    item: Item;
}

export function Draggable<Item = any>(props: DraggableProps<Item>) {
    const { item, children, ...divProps } = props;

    if (!children) return null;

    const { dragProps, isDragging } = useDrag({
        getItems: () => [
            {
                [CUSTOM_OBJECT_FORMAT]: JSON.stringify(item),
            },
        ],
    });

    return (
        <div {...dragProps} {...divProps}>
            {children({ isDragging })}
        </div>
    );
}
