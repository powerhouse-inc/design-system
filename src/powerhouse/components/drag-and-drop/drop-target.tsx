/* eslint-disable @typescript-eslint/no-misused-promises */
import { useRef } from 'react';
import { DropEvent, TextDropItem, useDrop } from 'react-aria';

import { CUSTOM_OBJECT_FORMAT } from './constants';

export interface DropTargetRenderProps {
    isDropTarget: boolean;
}

export interface DropTargetProps<Target = any, Item = any>
    extends Omit<React.HTMLAttributes<HTMLDivElement>, 'children'> {
    target: Target;
    children?: (props: DropTargetRenderProps) => React.ReactNode;
    onDropEvent: (item: Item, target: Target, event: DropEvent) => void;
}

export function DropTarget<Target = any, Item = any>(
    props: DropTargetProps<Target, Item>,
) {
    const { children, target, onDropEvent, ...divProps } = props;

    if (!children) return null;

    const ref = useRef(null);
    const { dropProps, isDropTarget } = useDrop({
        ref,
        async onDrop(e) {
            const item = e.items.find(
                item =>
                    item.kind === 'text' &&
                    item.types.has(CUSTOM_OBJECT_FORMAT),
            ) as TextDropItem | undefined;

            if (item) {
                const result = await item.getText(CUSTOM_OBJECT_FORMAT);
                onDropEvent(JSON.parse(result) as Item, target, e);
            }
        },
    });

    return (
        <div {...divProps} {...dropProps} ref={ref}>
            {children({ isDropTarget })}
        </div>
    );
}
