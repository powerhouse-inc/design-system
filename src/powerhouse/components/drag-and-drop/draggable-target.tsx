import { Draggable, DraggableRenderProps } from './draggable';
import {
    DropTarget,
    DropTargetProps,
    DropTargetRenderProps,
} from './drop-target';

export type DraggableTargetRenderProps = DraggableRenderProps &
    DropTargetRenderProps;

export interface DraggableTargetProps<Item = any> {
    item: Item;
    onDropEvent: DropTargetProps<Item, Item>['onDropEvent'];
    children: (props: DraggableTargetRenderProps) => React.ReactNode;
    dragNodeProps?: Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;
    targetNodeProps?: Omit<React.HTMLAttributes<HTMLDivElement>, 'children'>;
}

export function DraggableTarget<Item = any>(props: DraggableTargetProps<Item>) {
    const {
        item,
        onDropEvent,
        children,
        dragNodeProps = {},
        targetNodeProps = {},
    } = props;

    return (
        <DropTarget<Item, Item>
            onDropEvent={onDropEvent}
            target={item}
            {...targetNodeProps}
        >
            {({ isDropTarget }) => (
                <Draggable<Item> item={item} {...dragNodeProps}>
                    {({ isDragging }) => children({ isDropTarget, isDragging })}
                </Draggable>
            )}
        </DropTarget>
    );
}
