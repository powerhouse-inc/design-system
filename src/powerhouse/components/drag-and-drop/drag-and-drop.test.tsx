import { render, screen } from '@testing-library/react';

import { Draggable } from './draggable';
import { DraggableTarget } from './draggable-target';
import { DropTarget } from './drop-target';

describe('DragAndDrop Components', () => {
    describe('Draggable', () => {
        it('should match snapshot', () => {
            const { asFragment } = render(
                <Draggable item={{ id: 'item-1' }}>
                    {() => <div style={{ width: '100px', height: '100px' }} />}
                </Draggable>,
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it('should call children render prop function', () => {
            const children = jest.fn();

            render(<Draggable item={{ id: 'item-1' }}>{children}</Draggable>);

            expect(children).toHaveBeenCalled();
            expect(children).toHaveBeenCalledWith({ isDragging: false });
        });

        it('should inject props to the div wrapper element', () => {
            render(
                <Draggable item={{ id: 'item-1' }} data-testid="draggable-node">
                    {() => <div style={{ width: '100px', height: '100px' }} />}
                </Draggable>,
            );

            expect(screen.getByTestId('draggable-node')).toBeDefined();
        });
    });

    describe('DropTarget', () => {
        it('should match snapshot', () => {
            const onDropEvent = jest.fn();

            const { asFragment } = render(
                <DropTarget target={{ id: 'item-1' }} onDropEvent={onDropEvent}>
                    {() => <div style={{ width: '100px', height: '100px' }} />}
                </DropTarget>,
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it('should call children render prop function', () => {
            const children = jest.fn();
            const onDropEvent = jest.fn();

            render(
                <DropTarget target={{ id: 'item-1' }} onDropEvent={onDropEvent}>
                    {children}
                </DropTarget>,
            );

            expect(children).toHaveBeenCalled();
            expect(children).toHaveBeenCalledWith({ isDropTarget: false });
        });

        it('should inject props to the div wrapper element', () => {
            const onDropEvent = jest.fn();

            render(
                <DropTarget
                    target={{ id: 'item-1' }}
                    onDropEvent={onDropEvent}
                    data-testid="drop-target-node"
                >
                    {() => <div style={{ width: '100px', height: '100px' }} />}
                </DropTarget>,
            );

            expect(screen.getByTestId('drop-target-node')).toBeDefined();
        });
    });

    describe('DraggableTarget', () => {
        it('should match snapshot', () => {
            const { asFragment } = render(
                <DraggableTarget item={{ id: 'item-1' }} onDropEvent={() => {}}>
                    {() => <div style={{ width: '100px', height: '100px' }} />}
                </DraggableTarget>,
            );

            expect(asFragment()).toMatchSnapshot();
        });

        it('should call children render prop function', () => {
            const children = jest.fn();
            const onDropEvent = jest.fn();

            render(
                <DraggableTarget
                    item={{ id: 'item-1' }}
                    onDropEvent={onDropEvent}
                >
                    {children}
                </DraggableTarget>,
            );

            expect(children).toHaveBeenCalled();
            expect(children).toHaveBeenCalledWith({
                isDragging: false,
                isDropTarget: false,
            });
        });

        it('should inject props to the div wrapper elements', () => {
            const onDropEvent = jest.fn();

            render(
                <DraggableTarget
                    item={{ id: 'item-1' }}
                    onDropEvent={onDropEvent}
                    targetNodeProps={{ id: 'target-node', className: 'target' }}
                    dragNodeProps={{
                        id: 'draggable-node',
                        className: 'draggable',
                    }}
                >
                    {() => <div style={{ width: '100px', height: '100px' }} />}
                </DraggableTarget>,
            );

            const targetComponent = document.querySelector('#target-node');
            const draggableComponent =
                document.querySelector('#draggable-node');

            expect(targetComponent).toBeDefined();
            expect(targetComponent).toHaveClass('target');
            expect(draggableComponent).toBeDefined();
            expect(draggableComponent).toHaveClass('draggable');
        });
    });
});
