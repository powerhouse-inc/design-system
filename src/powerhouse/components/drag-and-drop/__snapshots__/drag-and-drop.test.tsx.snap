// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DragAndDrop Components Draggable should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-describedby="react-aria-description-0"
    draggable="true"
  >
    <div
      style="width: 100px; height: 100px;"
    />
  </div>
</DocumentFragment>
`;

exports[`DragAndDrop Components DraggableTarget should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      aria-describedby="react-aria-description-3"
      draggable="true"
    >
      <div
        style="width: 100px; height: 100px;"
      />
    </div>
  </div>
</DocumentFragment>
`;

exports[`DragAndDrop Components DropTarget should match snapshot 1`] = `
<DocumentFragment>
  <div>
    <div
      style="width: 100px; height: 100px;"
    />
  </div>
</DocumentFragment>
`;
